crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,311 @@
+use std::{
+    env, fs,
+    io::{self, BufReader, Cursor, ErrorKind, Read, Seek},
+    path::Path,
+};
+
+use semver::Version;
+use serde::Deserialize;
+use typst::{diag::bail, diag::StrResult, eval::eco_format};
+use xz2::bufread::XzDecoder;
+use zip::ZipArchive;
+
+use crate::args::UpdateCommand;
+
+const TYPST_GITHUB_ORG: &str = "typst";
+const TYPST_REPO: &str = "typst";
+
+/// Self update the typst CLI binary.
+///
+/// Fetches a target release or the latest release (if no version was specified)
+/// from GitHub, unpacks it and self replaces the current binary with the
+/// pre-compiled asset from the downloaded release.
+pub fn update(command: UpdateCommand) -> StrResult<()> {
+    match self_update_permitted()? {
+        SelfUpdatePermission::Deny => {
+            eprintln!("Self-update is disabled for this build of the typst cli");
+            eprintln!("You should probably use your system package manager");
+            bail!("update failed");
+        }
+        SelfUpdatePermission::Permit => {}
+    }
+
+    if let Some(ref version) = command.version {
+        let current_tag = env!("CARGO_PKG_VERSION").parse().unwrap();
+        if !command.force && version < &current_tag {
+            eprintln!("Certain downgraded Typst versions will not have the update command available");
```

Comment:
if we assume this ships in 0.8 (which is very probable), we could specifically check for 0.8 here maybe? and print the warning only then?

Reply:
Added.